test/test_framework.py

```diff
@@ -117,3 +126,52 @@ def test_analysis_runtime_opts(self):
         target_opts["figure_names"] = None
 
         self.assertEqual(analysis.options.__dict__, target_opts)
+
+    def test_after_job_fail(self):
+        """Verify that analysis is cancelled in case of job failure"""
+
+        class MyExp(BaseExperiment):
+            """Some arbitraty experiment"""
+
+            def __init__(self, qubits):
+                super().__init__(qubits)
+                self.analysis = MyAnalysis()
+
+            def circuits(self):
+                circ = QuantumCircuit(1, 1)
+                circ.measure(0, 0)
+                return [circ]
+
+        class MyAnalysis(BaseAnalysis):
+            """Analysis that is supposed to be cancelled, because of job failure"""
+
+            def _run_analysis(self, experiment_data):
+                res = AnalysisResultData(name="should not run", value="blaaaaaaa")
+                return [res], []
+
+        class MyBackend(AerSimulator):
+            """A backend that works with `MyJob`"""
+
+            def run(self, circuits, validate=False, parameter_binds=None, **run_options):
+                job = super().run(circuits, **run_options)
+                job.__class__ = MyJob
+                return job
+
+        class MyJob(AerJob):
+            """A job with status ERROR, that errors when the result is queried"""
+
+            def result(self, timeout=None):
+                raise QiskitError
+
+            def status(self):
+                return JobStatus.ERROR
+
+            def error_message(self):
+                """Job's error message"""
+                return "You're dealing with the wrong job, man"
+
+        backend = MyBackend()
+        exp = MyExp([0])
+        expdata = exp.run(backend=backend)
+        res = expdata.analysis_results()
+        self.assertEqual(len(res), 0)
```

Comment:
You can also validate `.analysis_status`.

Reply:
Done in https://github.com/Qiskit/qiskit-experiments/pull/871/commits/7dd4b77cd8787fc05ab79e93264f43a15ed722b9